openpgp/internal/ecc/curve25519.go

```diff
@@ -25,30 +23,66 @@ func (c *curve25519) GetCurveName() string {
 	return "curve25519"
 }
 
-func (c *curve25519) MarshalPoint(x, y *big.Int) []byte {
-	return x.Bytes()
+// MarshalBytePoint encodes the public point from native format, adding the prefix.
+// See https://datatracker.ietf.org/doc/html/draft-ietf-openpgp-crypto-refresh-06#section-5.5.5.6
+func (c *curve25519) MarshalBytePoint(point [] byte) []byte {
+	return append([]byte{0x40}, point...)
 }
 
-func (c *curve25519) UnmarshalPoint(point []byte) (x, y *big.Int) {
+// UnmarshalBytePoint decodes the public point to native format, removing the prefix.
+// See https://datatracker.ietf.org/doc/html/draft-ietf-openpgp-crypto-refresh-06#section-5.5.5.6
+func (c *curve25519) UnmarshalBytePoint(point []byte) []byte {
 	if len(point) != x25519lib.Size + 1 {
-		return nil, nil
+		return nil
 	}
 
-	x = new(big.Int)
-	x.SetBytes(point)
-	return x, nil
+	// Remove prefix
+	return point[1:]
 }
 
-func (c *curve25519) MarshalByteSecret(d []byte) []byte {
+// MarshalByteSecret encodes the secret scalar from native format.
+// Note that the EC secret scalar differs from the definition of public keys in
+// [Curve25519] in two ways: (1) the byte-ordering is big-endian, which is
+// more uniform with how big integers are represented in TLS, and (2) the
+// leading zeros are truncated.
+// See https://datatracker.ietf.org/doc/html/draft-ietf-openpgp-crypto-refresh-06#section-5.5.5.6.1.1
+// Here Leading zero bytes are stripped when encoding to MPI.
+func (c *curve25519) MarshalByteSecret(secret []byte) []byte {
+	// The following ensures that the private key is a number of the form
+	// 2^{254} + 8 * [0, 2^{251}), in order to avoid the small subgroup of
+	// the curve.
+	//
+	// This masking is done internally in the underlying lib and so is unnecessary
+	// for security, but OpenPGP implementations require that private keys be
+	// pre-masked.
+	//
+	// Do not copy, since changing the stored secret is an invariant for crypto
+	// operations
+	secret[0] &= 248
+	secret[31] &= 127
+	secret[31] |= 64
+
+	d := make([]byte, x25519lib.Size)
+	copyReversed(d, secret)
```

Comment:
I guess what you mean is, it doesn't matter if we modify the stored secret in this way, right? But we do make a copy, so the comment is a bit confusing, and anyway it seems cleaner for a marshal function not to modify the passed data, no? So I would do

```suggestion
	d := make([]byte, x25519lib.Size)
	copyReversed(d, secret)

	secret[0] &= 127
	secret[0] |= 64
	secret[31] &= 248
```

(and then maybe move around the comments or so)

Reply:
This will break the test that checks the Marshalling (ecdh.testMarshalUnmarshal) giving Curve25519 a different, special, behaviour (marshalled key vs original are different)

Reply:
Why? Doesn't it (or if not, can't you) test that after a marshal + unmarshal roundtrip, the parsed key is different from the original one? The test shouldn't rely on marshalling mutating the original key, imo

Reply:
> The test shouldn't rely on marshalling mutating the original key, imo
This is very true

On the other hand for all curves it's the same but curve25519

Reply:
Ah ok, I guess you mean the test checks that it's the same, not that it's different? Well yeah then we should add an exception to the test :sweat_smile: Because indeed we can't rely on that

Reply:
OK, will update this with the test then